Q24ECU/core/include/hal/hal_spi.h

```diff
@@ -0,0 +1,471 @@
+            Len--;
+            pRxBuffer++;
+        }
+    }
+}
+
+void SPI_CloseTransmission(SPI_Handle_t *pSPIHandle)
+{
+    pSPIHandle->pSPIx->CR2 &= ~(1 << SPI_CR2_TXEIE);
+    pSPIHandle->TxLen = 0;
+    pSPIHandle->pTxBuffer = NULL;
+    pSPIHandle->TxState = SPI_READY;
+}
+
+void SPI_CloseReception(SPI_Handle_t *pSPIHandle)
+{
+    pSPIHandle->pSPIx->CR2 &= ~(1 << SPI_CR2_RXNEIE);
+    pSPIHandle->RxLen = 0;
+    pSPIHandle->pRxBuffer = NULL;
+    pSPIHandle->RxState = SPI_READY;
+}
+
+__attribute__((weak)) void SPI_ApplicationEventCallback(SPI_Handle_t *pSPIHandle, uint8_t appEvent)
+{
+
+    // This is a weak implementation. The application may override this function.
+}
+
+static void spi_txe_interrupt_handle(SPI_Handle_t *pSPIHandle)
+{
+    // Check DFF in CR1
+    if (pSPIHandle->pSPIx->CR1 & (1 << SPI_CR1_DFF))
+    {
+        // 16 bit dff
+        pSPIHandle->pSPIx->DR = *((uint16_t *)pSPIHandle->pTxBuffer);
+        pSPIHandle->TxLen -= 2;
+        (uint16_t *)pSPIHandle->pTxBuffer++;
+    }
+    else
+    {
+        // 8 bit dff
+        pSPIHandle->pSPIx->DR = *pSPIHandle->pTxBuffer;
+        pSPIHandle->TxLen--;
+        pSPIHandle->pTxBuffer++;
+    }
+
+    if (!pSPIHandle->TxLen)
+        // TxLen is zero, so close the spi communication
+        // and inform app that TX is over
+        // this prevents interrupts from setting up of TXE flag
+        SPI_CloseTransmission(pSPIHandle);
+
+    SPI_ApplicationEventCallback(pSPIHandle, SPI_EVENT_TX_CMPLT);
+}
+
+static void spi_rxne_interrupt_handle(SPI_Handle_t *pSPIHandle)
+{
+    // Check DFF in CR1
+    if (pSPIHandle->pSPIx->CR1 & (1 << SPI_CR1_DFF))
+    {
+        // 16 bit dff
+        *((uint16_t *)pSPIHandle->pRxBuffer) = (uint16_t)pSPIHandle->pSPIx->DR;
+        pSPIHandle->RxLen -= 2;
+        (uint16_t *)pSPIHandle->pTxBuffer++;
+    }
+    else
+    {
+        // 8 bit dff
+        *(pSPIHandle->pRxBuffer) = (uint8_t)pSPIHandle->pSPIx->DR;
+        pSPIHandle->RxLen--;
+        pSPIHandle->pRxBuffer++;
+    }
+
+    if (!pSPIHandle->RxLen)
+    {
+        // TxLen is zero, so close the spi communication
+        // and inform app that TX is over
+        // this precents interrupts from setting up of TXE flag
+        SPI_CloseReception(pSPIHandle);
+
+        SPI_ApplicationEventCallback(pSPIHandle, SPI_EVENT_RX_CMPLT);
+    }
+}
+
+static void spi_ovr_err_interrupt_handle(SPI_Handle_t *pSPIHandle)
+{
+    uint8_t temp;
+    // 1. clear the OVR flag
+    if (pSPIHandle->TxState != SPI_BUSY_IN_TX)
+    {
+        temp = pSPIHandle->pSPIx->DR;
+        temp = pSPIHandle->pSPIx->SR;
+    }
+    (void)temp;
+    // If the SPI is busy in TX, then the error will not be cleared and the application will have to clear bit on own
+    // In this case, application should call SPI_ClearOVRFlag
+    // 2. inform the application
+    SPI_ApplicationEventCallback(pSPIHandle, SPI_EVENT_OVR_ERR);
+}
+
+void SPI_IRQHandling(SPI_Handle_t *pSPIHandle)
```

Comment:
Can we add some abstraction functions to enable/disable the interrupts? Refer to [lines 36-38](https://github.com/qfsae/zenith/blob/master/Q24ECU/core/src/interfaces/interface_uart.c) of the uart file

Reply:
Still needs to be done

Reply:
Fixed